### ISSUE TITLE
Use KGlobalAccel instead of lxqt-globalkeys

### ISSUE THREAD

**Issue by paulolieuthier** [lxqt/lxqt#507]
Now that KGlobalAccel's binary [is shipped with it](http://permalink.gmane.org/gmane.comp.kde.cvs/1384948), independent of Plasma, we should discuss moving to it and deprecating lxqt-globalkeys. AFAIK, using it would require only adding a `lxqt-kglobalaccel.desktop` in `/etc/xdg/autostart` with a line `X-LXQt-Module=true`.

According to @kuzmas, the two features KGlobalAccel misses compared to our implementation are: **making an arbitrary DBus call as an action** and **handling of multiple actions associated with one shortcut**.

**Comment by paulolieuthier** [comment]
My proposed solution doesn't work because KWin5 autostarts `kglobalaccel5` (so our module doesn't get started, because it is running already). If using KWin4, it will autostart `kglobalaccel` (the old one), which is compatible, apparently. If there is agreement, I will start porting our components to KGlobalAccel.

**Comment by paulolieuthier** [comment]
With Openbox the module is started correctly.

**Comment by jleclanche** [comment]
@paulolieuthier can backwards compatibility be maintained in some way? eg our components would work with both globalkeys and kglobalaccel

**Comment by jleclanche** [comment]
This is less important btw as long as @kuzmas is maintaining globalkeys, but I do think this is one of the more rewarding parts we can port to kf5.

**Comment by paulolieuthier** [comment]
@jleclanche I'm not sure how that could be done besides ifdefing it.

**Comment by dbluelle** [comment]
If I interpret https://projects.kde.org/projects/frameworks/kglobalaccel/repository/revisions/172c2a7813d60a0bd6174fd39c72aa2e106aa2fd correctly,
kglobalaccel is becoming a tier 3 kde framework which adds some additional kf5 dependencies  :worried: 
So maybe it's not such a good idea to replace lxqt-globalkeys?
( I remember reading somewhere that LXQt only wants to use tier 1 kde frameworks )

**Comment by paulolieuthier** [comment]
@dbluelle thanks, I didn't know of that. New dependencies: KConfig, KCoreAddons, KCrash, KDBusAddons, KI18n, KWindowSystem. That way it would be better not to have KGlobalAccel's binary after all and make lxqt-globalkeys compatible.

**Comment by paulolieuthier** [comment]
That's a lot of unnecessary dependencies. @jleclanche what do you think? KGlobalAccel's binary should be optional, right? That way we could depend on it for the library only and make lxqt-globalkeys compatible. @jleclanche what do you think?

**Comment by jleclanche** [comment]
Depending only on the lib/protocol and not the binary was the original intent when the binary was still in -runtime. Now it depends where they are willing to take this.

Ill think about it. But if it's possible to avoid the dep on the binary, for now, let's do it.

**Comment by paulolieuthier** [comment]
It's not possible as is. But we could contribute with upstream by making it optional in KGlobalAccel's CMakeLists.txts. Then we could distribute a package named lxqt-kglobalaccel, which would be KGlobalAccel compiled without the runtime. That's the best I can think of.

**Comment by PCMan** [comment]
I have read the source code of kdeglobalaccel stuff previously.
I'd say it's easier to improve our own global keys than removing kde
dependencies from the kde stuff.
As @kuzma is willing to maintain this part, my vote is don't touch it at
the moment.
If the KDE guys are willing to do it, that will be great.
If they don't, I'd say it's not a good idea to waste our precious man power
to do it right now.
We still have tons of important bugs to fix so this should really be a low
priority task.

**Comment by paulolieuthier** [comment]
@PCMan OK, you're right.

**Comment by kuzmas** [comment]
@PCMan I'm fine with maintaining lxqt-globalkeys.

But I remembered one more thing to consider: lxqt-globalkeys does not support wayland yet. Though later we can ~~steal~~ copy the code from somewhere ;-)

**Comment by paulolieuthier** [comment]
Just to make it clear: I agree we have little man power and need to focus it on what is more important and urgent. But I also think making lxqt components depend on KF5 libraries reduces the need for man power in the long term.

**Comment by agaida** [comment]
@paulolieuthier you're right, but please limit the KF5 dependencies to Tier 1

**Comment by PCMan** [comment]
It's clear that we can benefit from that. There are some issues, though.
1. It's still tier 3 now. We need to wait KDE guys for making it tier 1, if
possible.
2. Wayland support for it is a long way to go. So, we'll not get there now.
3. It's not well-tested. KDE is not fully depolyed with KF5 yet. After KFS
has been used for a while, potential bugs will be fixes and it will be
safer to widely adopt KF5 at that time. (So, not now)
4. KF5 is not yet available in some distros, such as Debian.
I support the idea of using tier 1 KF5 stuff, but it's not such an urgent
task.
Too early adoption can introduce more bugs than benefit since wayland
support is not ready. So we basically gain nothing.

**Comment by agaida** [comment]
KF5 isn't in sid nor experimental, thats true - but its in the new queue right now (KF5 5.6), so it is only a matter of time. That will affect LXQt not so much because only libqtxdg is uploaded right now and in the new queue too :)

**Comment by heliocastro** [comment]
Yes, @paulolieuthier is right, keep on Tier 1.

Just not base the decision on "is not done for prime time", "Debian not
released" or "Is not ready for wayland" or any possible vague excuse.
We should avoid create flamewars or throw biased opinions,

I suggest be more concise and do an logic approach:

1 - Is the new <LIBRARY> fits on LXQt needs as a replacement or new add ?
If yes, then
2 - Is it attends the basic necessities and lean requirements of LXQt ?
If yes, then
3 - Is it ready to test, even as in development ?
If Yes
4 - Need some work on it on upstream ?
If Yes, can we help ?
5 - Need some work to be in LXQt ?
If yes, try to talk upstream or back to item 2 and say No, or put a wait
condition, based on some time.

If all of this conditions can be satisfied, the i think is a good
candidate.
I'm talking on every possible candidate library or resource, not KDE ones
only.

Other reasons is just biased ones to prevent we move forward.

**Comment by jleclanche** [comment]
To be clear, I want us to move these things forward so that we have less workload and less to maintain. But at this point, kglobalaccel still pulls in too many unnecessary dependencies to be useful to us.

Looking at the code, the KConfig dependency could be easily dropped - it's used in two places, extremely lightly and only in the runtime.

KCoreAddons is used only for KAboutData. Seeing as the runtime has no way of exposing that data, it seems stupid that it does and it can be dropped.

The KI18n dependency is used for the god damn about data. 

The KCrash dependency is here only to set an "auto restart" flag on the app. It's worth looking at whether we can move this out of kcrash somehow, because that is a feature we need but currently expose in liblxqt.

The KDBusAddons dependency is for [KDBusService](http://api.kde.org/frameworks-api/frameworks5-apidocs/kdbusaddons/html/classKDBusService.html) only. It's light, so whatever... suggestions welcome on this one I guess.

The app links to KCoreAddons but doesn't actually seem to use it for anything. This can probably be dropped.

That leaves KWindowSystem which is already in use in lxqt.

**Comment by jleclanche** [comment]
Sent a patch to drop KCoreAddons and KI18n: https://git.reviewboard.kde.org/r/122680/

**Comment by PCMan** [comment]
On Mon, Feb 23, 2015 at 1:14 AM, Helio Chissini de Castro <
notifications@github.com> wrote:

> Yes, @paulolieuthier is right, keep on Tier 1.
>
> Just not base the decision on "is not done for prime time", "Debian not
> released" or "Is not ready for wayland" or any possible vague excuse.
> We should avoid create flamewars or throw biased opinions,
>
> These are not excuses, nor biased.
We have 167 open issues in the tracker now and we have only 5-10 active
devs.
Most of the issues are not yet worked on.
Yes, we definitely need to move forward, but things need to be prioritized.
That's my point. I'm not against moving forward.
When we move to Qt5, we encountered tons of regression bugs.
Then I did workarounds here and there and help fixed the Qt 5 issue.
While introducing KWindowsystem, we again got some problems.
Luckily it's fixed now thanks to other developers and its upstream.
While moving forward, it's good to be more conservative.
We still have many unresolved bugs.
If introducing a new library can fix some of them, then do it.
Otherwise, it might be a better strategy to fix existing problems first and
not complicate the problems with new lib at the moment.
I support the proposal of kdeglobalaccel, but it's not so urgent.
While we are fixing our exsting bugs, upstream developers can improve KF5
at the same time.
Then when we have the time to work on it, KF5 already becomes more polished
and better tested.
So don't get me wrong.

**Comment by palinek** [comment]
Note: I've submitted patch to KGlobalAccel https://phabricator.kde.org/D5521 to allow removing dependencies

**Comment by tsujan** [comment]
I think this can be closed, at least for the following reasons:

 * KGlobalAccel has remained dependent on what we don't need.
 * KDE hasn't moved in the direction of modularity, and there's no sing that it will.
 * We now support Wayland and know that KGlobalAccel isn't the solution. There's no global implementation, and each Wayland compositor has its own way of defining shortcuts.
